### PR TITLE
Feature/auth

### LINKED ISSUE #5: [FEATURE] OAuth Authentication
### Type of feature

Something new

### The name of the feature

OAuth Authentication

### Feature description

Create the OAuth Authentication Golang Side with the deep link for the flutter side

### Additional information

- [ ] Create the user model
- [ ] Create routes
- [ ] Golang Oauth 
- [ ] Create Flutter Deep Link

### PR DESCRIPTION
**Pull request by Spacelocust**
## Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Closes #5

## Type of change

Please add any relevant information to the type of change you've made.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update (or is a documentation update)
-   [ ] HIGH IMPORTANCE (requires a fast merging)

## Has This Been Tested?

Please describe the tests that you ran to verify your changes. It is way more likely for a PR to be merged and approved with successful tests.

-   [x] Yes, I ran tests
-   [ ] No, I did not include any test
-   [ ] No test is needed for this change

### CODE REVIEW

File: api/internal/oauth/config.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	"os"
+
+	"github.com/gorilla/sessions"
+	"github.com/markbates/goth"
+	"github.com/markbates/goth/gothic"
+	"github.com/markbates/goth/providers/steam"
+)
+
+// TODO: Just for testing purposes
+var baseUrl = "http://10.0.2.2:5000"
```
Comment (YummYume) [inline api/internal/oauth/config.go]: Retirer ?

File: api/internal/server/oauth.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+func (s *Server) OAuth(c *gin.Context) {
+	oauth.BeginAuthHandler(c)
+}
+
+const pageLoading = `<!DOCTYPE html>
+<html lang="en">
+<head>
+		<meta charset="UTF-8">
+		<meta name="viewport" content="width=device-width, initial-scale=1.0">
+		<title>Centered Image</title>
```
Comment (YummYume) [inline api/internal/server/oauth.go]:
```suggestion
		<title>For Democracy - Login</title>
		<meta name="robots" content="noindex, nofollow">
```

File: mobile/lib/l10n/app_en.arb [hunk 3]
```diff
@@ -1,131 +1,131 @@
...
-    "statisticMissionTime": "Mission time",
-    "statisticBugKills": "Bug kills",
-    "statisticAutomatonKills": "Automaton kills",
-    "statisticIlluminateKills": "Illuminate kills",
-    "statisticBulletsFired": "Bullets fired",
-    "statisticBulletsHit": "Bullets hit",
-    "statisticTimePlayed": "Time played",
-    "statisticDeaths": "Deaths",
-    "statisticRevives": "Revives"
+  "@@locale": "en",
```
Comment (YummYume) [inline mobile/lib/l10n/app_en.arb]: Remettre comme avant sinon on va avoir des conflits inutiles

File: mobile/lib/screens/web_oauth_screen.dart [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import 'package:mobile/services/api_service.dart';
+import 'package:mobile/services/oauth_service.dart';
+import 'package:mobile/services/token_service.dart';
+import 'package:mobile/states/auth_state.dart';
+import 'package:provider/provider.dart';
+import 'package:webview_cookie_manager/webview_cookie_manager.dart';
+import 'package:webview_flutter/webview_flutter.dart';
+import 'package:webview_flutter_android/webview_flutter_android.dart';
+
+class WebOauthScreen extends StatefulWidget {
```
Comment (YummYume) [inline mobile/lib/screens/web_oauth_screen.dart]:
```suggestion
class WebOAuthScreen extends StatefulWidget {
```

File: mobile/lib/screens/web_oauth_screen.dart [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  const WebOauthScreen({super.key});
+
+  @override
+  State<WebOauthScreen> createState() => _WebOauthScreenState();
+}
+
+class _WebOauthScreenState extends State<WebOauthScreen> {
```
Comment (YummYume) [inline mobile/lib/screens/web_oauth_screen.dart]:
```suggestion
class _WebOAuthScreenState extends State<WebOauthScreen> {
```

File: mobile/lib/screens/web_oauth_screen.dart [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  late final WebViewController _controller;
+  final cookieManager = WebviewCookieManager();
+
+  String baseUrl = dotenv.get(APIService.baseUrlEnv);
+  String steamOauthUrl = '/oauth/steam';
+  String steamCallbackUrl = '/oauth/steam/callback';
```
Comment (YummYume) [inline mobile/lib/screens/web_oauth_screen.dart]: const ?

File: mobile/lib/screens/web_oauth_screen.dart [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                  );
+                }
+              } finally {
+                await cookieManager.clearCookies();
+              }
+
+              if (mounted) {
+                Navigator.pop(context);
+              }
+            }
```
Comment (YummYume) [inline mobile/lib/screens/web_oauth_screen.dart]: On comprend pas trop

File: mobile/lib/screens/web_oauth_screen.dart [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      child: WebViewWidget(controller: _controller),
+    );
+  }
+
+  // Retrieve the cookie value from the WebView
+  Future<String?> getCookie(String url, String name) async {
+    final cookies = await cookieManager.getCookies(url);
+
+    final Cookie cookie = cookies.firstWhere((cookie) => cookie.name == name,
+        orElse: () => Cookie(name, ""));
```
Comment (YummYume) [inline mobile/lib/screens/web_oauth_screen.dart]:
```suggestion
        orElse: () => Cookie(name, ""),);
```

File: mobile/lib/services/api_service.dart [hunk 9]
```diff
@@ -9,9 +12,40 @@ abstract class APIService {
...
+    },
+    onResponse: (response, handler) {
+      return handler.next(response);
+    },
+    onError: (e, handler) async {
+      if (e.response != null) {
+        if (e.response!.statusCode == 401 &&
+            e.response!.statusMessage != null) {
+          if (e.response!.statusMessage!.contains('token expired') ||
+              e.response!.statusMessage!.contains('token invalid')) {
```
Comment (YummYume) [inline mobile/lib/services/api_service.dart:42]: Un peu bizarre de vérifier un message

File: mobile/lib/services/oauth_service.dart [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import 'package:mobile/models/user.dart';
+import 'package:mobile/services/api_service.dart';
+
+abstract class OauthService {
+  static const String oauthUrl = '/oauth/me';
```
Comment (YummYume) [inline mobile/lib/services/oauth_service.dart]:
```suggestion
abstract class OAuthService {
  static const String oAuthUrl = '/oauth/me';
```

File: mobile/lib/widgets/components/square_avatar.dart [hunk 11]
```diff
@@ -0,0 +1,26 @@
+import 'package:flutter/material.dart';
+
+/// Square avatar widget.
+class SquareAvatar extends StatelessWidget {
+  final Image avatar;
+  final void Function()? onTap;
+  final double size;
```
Comment (YummYume) [inline mobile/lib/widgets/components/square_avatar.dart]: Espacement

File: mobile/lib/widgets/components/square_avatar.dart [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        onTap: onTap,
+        child: DecoratedBox(
+          decoration: BoxDecoration(
+            image: DecorationImage(
+              image: avatar.image,
+            ),
+            borderRadius: BorderRadius.circular(5.0),
+          ),
+          child: SizedBox(width: size, height: size),
+        ));
```
Comment (YummYume) [inline mobile/lib/widgets/components/square_avatar.dart]:
```suggestion
  const SquareAvatar(
      {super.key, required this.avatar, this.size = 40, this.onTap,});

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
        onTap: onTap,
        child: DecoratedBox(
          decoration: BoxDecoration(
            image: DecorationImage(
              image: avatar.image,
            ),
            borderRadius: BorderRadius.circular(5.0),
          ),
          child: SizedBox(width: size, height: size),
        ),);
```

File: mobile/lib/widgets/layout/main_navigation_drawer.dart [hunk 13]
```diff
@@ -59,11 +65,106 @@ class MainNavigationDrawer extends StatelessWidget {
...
+          const SizedBox(width: 16),
+          Text(
+            user.username,
+            style: Theme.of(context).textTheme.titleLarge,
+          ),
+        ],
+      );
+    } else {
+      return Container();
+    }
```
Comment (YummYume) [inline mobile/lib/widgets/layout/main_navigation_drawer.dart:116]:
```suggestion
    }
    
    return Container();
```

On est sûr de vouloir mettre un container vide ?

File: mobile/lib/widgets/layout/main_navigation_drawer.dart [hunk 14]
```diff
@@ -59,11 +65,106 @@ class MainNavigationDrawer extends StatelessWidget {
...
+    if (context.watch<AuthState>().user != null) {
+      return ElevatedButton(
+          onPressed: () {
+            context.read<AuthState>().setUser(null);
+            SecureStorageService().deleteSecureData("token");
+          },
+          child: Text(
+            AppLocalizations.of(context)!.logout,
+            style: Theme.of(context).textTheme.titleSmall,
+          ));
```
Comment (YummYume) [inline mobile/lib/widgets/layout/main_navigation_drawer.dart]:
```suggestion
          ),);
```

File: mobile/lib/widgets/layout/main_navigation_drawer.dart [hunk 15]
```diff
@@ -59,11 +65,106 @@ class MainNavigationDrawer extends StatelessWidget {
...
+    } else {
```
Comment (YummYume) [inline mobile/lib/widgets/layout/main_navigation_drawer.dart]:
```suggestion
    }
```

File: mobile/lib/widgets/layout/main_navigation_drawer.dart [hunk 16]
```diff
@@ -59,11 +65,106 @@ class MainNavigationDrawer extends StatelessWidget {
...
+      return ElevatedButton(
+          style: ElevatedButton.styleFrom(backgroundColor: Colors.black),
+          onPressed: openSteamModal,
+          child: const Padding(
+              padding: EdgeInsets.all(10),
+              child: Image(
+                image: AssetImage('assets/steam-logo.png'),
+                height: 30,
+                alignment: Alignment.centerLeft,
+              )));
```
Comment (YummYume) [inline mobile/lib/widgets/layout/main_navigation_drawer.dart]:
```suggestion
              ),),);
```

File: mobile/lib/widgets/layout/main_scaffold.dart [hunk 17]
```diff
@@ -25,20 +25,37 @@ class MainScaffold extends StatelessWidget {
...
+    final user = context.watch<AuthState>().user;
+
+    if (user != null) {
+      return Padding(
+        padding: const EdgeInsets.only(right: 10.0),
+        child: CircleAvatar(
+          backgroundImage: NetworkImage(user.avatarUrl),
+        ),
+      );
+    } else {
```
Comment (YummYume) [inline mobile/lib/widgets/layout/main_scaffold.dart]:
```suggestion
    }
```

File: mobile/lib/widgets/layout/main_scaffold.dart [hunk 18]
```diff
@@ -25,20 +25,37 @@ class MainScaffold extends StatelessWidget {
...
+      return const Padding(
+          padding: EdgeInsets.only(right: 10.0),
+          child: Icon(Icons.account_circle));
```
Comment (YummYume) [inline mobile/lib/widgets/layout/main_scaffold.dart]:
```suggestion
          child: Icon(Icons.account_circle),);
```

File: mobile/pubspec.yaml [hunk 19]
```diff
@@ -72,6 +77,7 @@ flutter:
 
   # To add assets to your application, add an assets section, like this:
   assets:
+    - assets/
```
Comment (YummYume) [inline mobile/pubspec.yaml:80]: ça include tout automatiquement ?

**Review by YummYume — CHANGES_REQUESTED** [review]

File: mobile/pubspec.yaml (hunk 19 above)
Comment (Spacelocust) [reply mobile/pubspec.yaml:80]: Ouaip

File: mobile/lib/screens/web_oauth_screen.dart (hunk 7 above)
Comment (Spacelocust) [reply mobile/lib/screens/web_oauth_screen.dart]: Je vais essayer d'améliorer et de commenter

File: mobile/lib/services/api_service.dart (hunk 9 above)
Comment (Spacelocust) [reply mobile/lib/services/api_service.dart:42]: Je n'ai pas réellement d'autres moyens

File: mobile/lib/widgets/layout/main_navigation_drawer.dart (hunk 13 above)
Comment (Spacelocust) [reply mobile/lib/widgets/layout/main_navigation_drawer.dart:116]: J'aurais préféré mettre rien ou un Widget qui fait office de vide

File: api/internal/server/oauth.go [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+func (s *Server) OAuth(c *gin.Context) {
+	oauth.BeginAuthHandler(c)
+}
+
+const pageLoading = `<!DOCTYPE html>
+<html lang="en">
+<head>
+		<meta charset="UTF-8">
+		<meta name="viewport" content="width=device-width, initial-scale=1.0">
+		<title>For Democracy - Login</title>
```
Comment (YummYume) [inline api/internal/server/oauth.go:144]:
```suggestion
		<title>For Democracy - Login</title>
		<meta name="robots" content="noindex, nofollow">
```

**Review by YummYume — APPROVED** [review]

**Pull request merged**